reverse dns streaming processor

ssoroka opened the pull request:
Resolves https://github.com/influxdata/telegraf/issues/7245

- adds support for streaming processors
- adds reverse dns lookup streaming processor with rdns cache

danielnelson commented on models/running_processor.go:
```diff
@@ -71,34 +73,79 @@ func (r *RunningProcessor) Init() error {
 
-	ret := []telegraf.Metric{}
+	r.Config.Filter.Modify(metric)
+	if len(metric.FieldList()) == 0 {
```
If all fields are removed then the metric needs to be dropped, I believe we are passing the metric.

danielnelson commented on models/running_processor.go:
```diff
@@ -71,34 +73,79 @@ func (r *RunningProcessor) Init() error {
+		if sa.acc.IsStreamClosed() {
+			return nil
+		}
+		goto retry
```
This isn't a very reasonable spot to use goto, just use a for loop.

danielnelson commented on accumulator.go:
```diff
@@ -83,3 +83,32 @@ type TrackingAccumulator interface {
 	Delivered() <-chan DeliveryInfo
 }
+
+// StreamingAccumulator provides a way to interact with the metric stream,
+// both being notified of new metrics (GetNextMetric), and being able to pass
+// them on downstream (PassMetric).
+// (presumably after you've changed it)
+type StreamingAccumulator interface {
```
Definitely rename because it isn't an Accumulator.

danielnelson commented on plugins/processors/reversedns/reversedns.go:
```diff
@@ -0,0 +1,193 @@
+	defer r.exitWg.Done()
+
+	for futureMetricFunc := range outChannel {
+		acc.PassMetric(futureMetricFunc())
```
I'm confused here, all metrics are processed one at a time?  I'm okay with this as we now have every processor running concurrently and there is a tiny buffered chan between each of them, but it seems like we could do all of this with the existing Processor interface and some small changes to the RunningProcessor.

Let's talk offline about how much concurrency we need.  We don't need it at on every axis and limiting it will simplify things quite a bit.

ssoroka replied:
🤦 no just me being stupid. fixed will push.

danielnelson commented on plugins/processors/reversedns/reversedns.go:
```diff
@@ -0,0 +1,193 @@
+
+const sampleConfig = `
+[[processors.reversedns]]
+	# For optimal performance, you may want to limit which metrics are passed to this
```
2 space indent

danielnelson commented on plugins/processors/reversedns/reversedns.go:
```diff
@@ -0,0 +1,193 @@
+// }
+
+func init() {
+	processors.AddStreaming("reversedns", func() telegraf.StreamingProcessor {
```
This file is not imported in all.go, so its not registered.

danielnelson commented on plugins/processors/reversedns/reversedns.go:
```diff
@@ -0,0 +1,193 @@
+	reverseDNSCache *ReverseDNSCache
+
+	Lookups            []lookupEntry `toml:"lookup"`
+	CacheTTL           time.Duration `toml:"cache_ttl"`
+	LookupTimeout      time.Duration `toml:"lookup_timeout"`
```
These can't be unmarshaled from TOML

ssoroka replied:
I thought we fixed that. Will test.

ssoroka commented on agent/agent.go:
```diff
@@ -349,31 +348,49 @@ func (a *Agent) gatherOnce(
-	for metric := range src {
-		metrics := a.applyProcessors(metric)
+	wg := sync.WaitGroup{}
+	in := src
+	out := dest
+	for i, processor := range a.Config.Processors {
+		wg.Add(1)
+		nextCh := make(chan telegraf.Metric, 10)
```
Going primarily for clarity here. Not sure what to change at the moment.

ssoroka commented on agent/streamingaccumulator.go:
```diff
@@ -0,0 +1,91 @@
+
+// GetNextMetric returns a metric or blocks until one is available.
+func (sa *streamingAccumulator) GetNextMetric() telegraf.Metric {
+	sa.Lock()
```
can be called in parallel.

ssoroka commented on config/config.go:
```diff
@@ -925,11 +920,18 @@ func (c *Config) addAggregator(name string, table *ast.Table) error {
 		return fmt.Errorf("Undefined but requested processor: %s", name)
 	}
-	processor := creator()
+	var processor telegraf.PluginDescriber
```
at the moment they don't share any other functions. Hmm. this might be possible with the new wrapper. I'll look at that.

ssoroka commented on models/running_processor.go:
```diff
@@ -9,8 +10,9 @@ import (
-	Processor telegraf.Processor
+	log telegraf.Logger
+	// Processor can be a Processor or a StreamingProcessor
+	Processor telegraf.PluginDescriber
```
Hmm. I'm thinking I might just be able to make it always a streaming processor now, since I can wrap the other.

ssoroka commented on processor.go:
```diff
@@ -1,12 +1,82 @@
+}
+
+// NewProcessorWrapper turns a standard processor into a streaming processor
+func NewProcessorWrapper(p Processor) StreamingProcessor {
```
Good feedback. It needs a better name. It turns a Processor into a StreamingProcessor (by wrapping it, but that's less important). A descriptive name might be `NewStreamingProcessorFromProcessor` ?

ssoroka commented on plugins/processors/reversedns/reversedns.go:
```diff
@@ -5,46 +5,47 @@ import (
 )
 
 const sampleConfig = `
 [[processors.reversedns]]
```
remove this

ssoroka commented on plugins/processors/reversedns/reversedns.go:
```diff
@@ -58,26 +59,26 @@ type ReverseDNS struct {
-	LookupTimeout      time.Duration `toml:"lookup_timeout"`
-	MaxParallelLookups int           `toml:"max_parallel_lookups"`
+	Lookups            []lookupEntry     `toml:"lookup"`
+	CacheTTL           internal.Duration `toml:"cache_ttl"`
```
use config.Duration

danielnelson commented on models/running_processor.go:
```diff
@@ -71,34 +71,43 @@ func (r *RunningProcessor) Init() error {
+		if filteredMetric == nil {
+			acc.PassMetric(m)
 			continue
 		}
-
-		rp.Config.Filter.Modify(metric)
-		if len(metric.FieldList()) == 0 {
-			rp.metricFiltered(metric)
+		if len(m.FieldList()) == 0 {
+			acc.DropMetric(filteredMetric)
```
Would be best to merge these functions again so you don't need to double check the number of fields.  Either that or return multiple values, but I'm not sure I see the advantage of the change.  Make sure the metric is properly dropped (must call metric.Drop()).

ssoroka commented on plugins/processors/reversedns/parallel/parallel.go:
```diff
@@ -0,0 +1,83 @@
+		if m != nil {
+			p.acc.PassMetric(m)
+		}
+		p.Done()
```
That pattern works for waiting for an async function to finish. this is confirming receipt of the metric, so the add and done follows the metric-start and metric-completion.

ssoroka commented on plugins/processors/reversedns/parallel/parallel.go:
```diff
@@ -0,0 +1,83 @@
+	o.future = make(futureMetric)
+	go o.fn(o)
+	return o.future
+}
```
I've updated this to support multiple metrics, still ordered, without deferring to slices.

ssoroka commented on plugins/processors/reversedns/parallel/parallel.go:
```diff
@@ -0,0 +1,83 @@
+	return p
+}
+
+func (p *Parallel) Unordered() *Parallel {
```
resolved this. Split it into two implementations to make sure I could resolve it properly. No issues.

danielnelson commented on agent/agent.go:
```diff
@@ -94,7 +93,7 @@ func (a *Agent) Run(ctx context.Context) error {
 			defer wg.Done()
 
-			err := a.runProcessors(src, dst)
+			err := a.runProcessors(src, dst, a.Config.Processors)
```
To fix this, we need to call Start() earlier, before the pipeline and any plugins are run.  We also need to tear down everything on error.

On error, call Stop if and only if Start returned without any error.

ssoroka replied:
Inputs don't write to dst. This processor is the only thing that writes to dst.

danielnelson replied:
Oh yeah, well no panic but the pipeline is still broken and will hang, so I think the solution is the same.

ssoroka replied:
If the processor fails to start, Telegraf is borked and won't run anyway, and we're fine to exit the process with an error, aren't we? I would actually think that we should be returning an error on line 98-99 and exiting immediately. Does Telegraf try to keep running in this situation? Keep in mind runProcessors _only_ returns an error on startup failures.

ssoroka commented:
addressed feedback:
- fixed serializers/parsers
- fixed ApplyFilters
- switch parallel.Ordered to use workers
- added parallel.Ordered test
- add ordered benchmark; performance is good.
- use semaphores instead of introducing new pool concept.

ssoroka commented on agent/agent.go:
```diff
@@ -573,39 +572,50 @@ func (a *Agent) runAggregators(
 
 			if !dropOriginal {
-				dst <- metric // keep original.
+				unit.outputC <- metric // keep original.
```
Is this so we can also use aggregators to filter out the original metric?

ssoroka commented on plugins/processors/reversedns/parallel/unordered.go:
```diff
@@ -31,18 +31,18 @@ func NewUnordered(acc telegraf.MetricStreamAccumulator, workerCount int64) *Unor
 	p.sem.Acquire(context.TODO(), 1)
 	go func() {
-		fn(p)
+		fn(p.acc)
```
This is a breaking change. The previous version used the idea that Unordered matched the accumulator interface so it could write to queue.

ssoroka commented:
closing in favor of https://github.com/influxdata/telegraf/pull/7639
